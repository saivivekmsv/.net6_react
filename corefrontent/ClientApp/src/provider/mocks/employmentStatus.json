{
  "data": [
    {
      "label": "Active",
      "value": 1
    },
    {
      "label": "Retired",
      "value": 2
    },
    {
      "label": "Terminated",
      "value": 3
    },
    {
      "label": "Leave",
      "value": 4
    },
    {
      "label": "Military Leave",
      "value": 5
    },
    {
      "label": "Deceased",
      "value": 6
    },
    {
      "label": "Disabled",
      "value": 7
    },
    {
      "label": "Others",
      "value": 8
    }
  ]
}
