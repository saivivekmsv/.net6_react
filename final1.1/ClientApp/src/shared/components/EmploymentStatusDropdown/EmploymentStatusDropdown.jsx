import { find, get, isEmpty, toLower } from "lodash";
import React, { useState } from "react";
import { Form, Button } from "react-bootstrap";
import { SearchableList, FormControl } from "../";
import { useDeepEffect, useRequest } from "../../abstracts";
import {
  addMasterEmploymentStatus,
  checkMasterEmploymentStatusExists,
  getMasterEmploymentStatuses,
} from "../../services";
import { errors, tranformListToDropdownValues } from "../../utils";

const EmploymentStatusDropdown = ({
  data,
  onSelect,
  companyId,
  setData,
  value,
}) => {
  const [newEmploymentStatus, setNewEmploymentStatus] = useState("");
  const [request, setRequest] = useState({
    method: getMasterEmploymentStatuses,
    payload: companyId,
    stopTrigger: !isEmpty(data),
  });
  const {
    loading: checkingMasterStatusName,
    response: masterStatusCheckResponse,
  } = useRequest({
    method: checkMasterEmploymentStatusExists,
    payload: newEmploymentStatus,
    triggerOnlyOnUpdate: true,
    defaultResponse: true,
  });
  const [lengthCheck, setLengthCheck] = useState(true);
  const format = /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]+/;
  const { response, loading } = useRequest(request);
  useDeepEffect(
    () => {
      if (!loading) {
        if (newEmploymentStatus) {
          const filteredValue = find(
            response,
            (item) => toLower(item.name) === toLower(newEmploymentStatus)
          );
          onSelect(
            get(filteredValue, "id", ""),
            get(filteredValue, "name", "")
          );
        }
        setData(tranformListToDropdownValues(response || [], "name", "id"));
      }
    },
    [loading],
    true
  );

  const newEmploymentStatusChange = (e) => {
    const elem = e.target;
    const val = e.target.value;
    elem.value = val;
    setLengthCheck(val.length <= 50);
    setNewEmploymentStatus(val);
  };

  const onNewEmploymentStatusChangeAdd = () => {
    if (
      !masterStatusCheckResponse &&
      lengthCheck &&
      !format.test(newEmploymentStatus && newEmploymentStatus)
    ) {
      setRequest({
        ...request,
        stopTrigger: false,
        method: addMasterEmploymentStatus,
        payload: {
          name: newEmploymentStatus,
          id: companyId,
        },
      });
    }
  };
  return (
    <div className="dropdside-custom-dropdown">
      <div>
        <SearchableList
          label="Select an Employment Status or create a new one"
          options={data}
          onSelect={onSelect}
          selectedValue={value}
          // isNotTypeAhead
        />
      </div>
      <div className="line-separator"></div>
      <FormControl
        label="Add a new Employment Status type :"
        hasSuggestion
        isSuggestionLoading={checkingMasterStatusName}
        isValidSuggestion={
          !masterStatusCheckResponse &&
          lengthCheck &&
          !format.test(newEmploymentStatus)
        }
        suggestionErrorMessage={
          !lengthCheck
            ? errors.masterClassificationName.length
            : format.test(newEmploymentStatus)
            ? errors.masterEmploymentStatus.characters
            : errors.masterEmploymentStatus.alreadyExist
        }
      >
        <input
          className="searchable-list-input-text"
          name="newEmploymentStatusType"
          type="text"
          size="md"
          onChange={newEmploymentStatusChange}
          onKeyDown={(e) => {
            if (e.key == " ") {
              e.preventDefault();
              e.target.value += " ";
            }
          }}
        />
      </FormControl>
      <Form.Group>
        <Button type="button" onClick={onNewEmploymentStatusChangeAdd}>
          Add
        </Button>
      </Form.Group>
    </div>
  );
};

export default EmploymentStatusDropdown;
