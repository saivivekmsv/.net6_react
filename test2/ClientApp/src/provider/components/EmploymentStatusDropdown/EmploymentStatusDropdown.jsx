import { find, get, isEmpty, toLower } from "lodash";
import React, { useState } from "react";
import { Form, Button } from "react-bootstrap";
import { Field, Formik } from "formik";
import * as Yup from "yup";
import { SearchableList, LoaderWrapper } from "../";
import { useDeepEffect, useRequest } from "../../abstracts";
import {
  addMasterEmploymentStatus,
  getMasterEmploymentStatuses,
} from "../../services";
import { FieldInput } from "../../components";
import { errors, tranformListToDropdownValues } from "../../utils";

const EmploymentStatusDropdown = ({
  data,
  onSelect,
  companyId,
  setData,
  value,
}) => {
  const [newEmploymentStatus, setNewEmploymentStatus] = useState();
  const [request, setRequest] = useState({
    method: getMasterEmploymentStatuses,
    payload: companyId,
    stopTrigger: !isEmpty(data),
  });
  const { response, loading } = useRequest(request);

  const schema = Yup.object().shape({
    newEmploymentStatusType: Yup.string()
      .required(errors.masterEmploymentStatus.required)
      .max(50, errors.masterEmploymentStatus.length)
      .test(
        "newEmploymentStatusType",
        errors.masterEmploymentStatus.characters,
        (val) => !checkSplChar(val)
      )
      .test(
        "newEmploymentStatusType",
        errors.masterEmploymentStatus.alreadyExist,
        (val) => !alreadyExist(val)
      ),
  });

  const checkSplChar = (val) => {
    const format = /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]+/;
    if (format.test(val) === true) return true;
    else return false;
  };
  const alreadyExist = (val) => {
    const temp = data.map((item) => {
      if (toLower(item.label) === toLower(val)) return true;
      else return false;
    });
    return temp.includes(true);
  };

  useDeepEffect(
    () => {
      if (!loading) {
        if (newEmploymentStatus) {
          const filteredValue = find(
            response,
            (item) => toLower(item.name) === toLower(newEmploymentStatus)
          );
          onSelect(
            get(filteredValue, "id", ""),
            get(filteredValue, "name", "")
          );
        }
        setData(tranformListToDropdownValues(response || [], "name", "id"));
      }
    },
    [loading],
    true
  );
  const keyDown = (e) => {
    if (e.keyCode == 32) {
      e.preventDefault();
      e.target.value += " ";
    }
  };
  const onFormSubmit = (values) => {
    setNewEmploymentStatus(values["newEmploymentStatusType"]);
    setRequest({
      ...request,
      stopTrigger: false,
      method: addMasterEmploymentStatus,
      payload: {
        name: values["newEmploymentStatusType"],
        id: companyId,
      },
    });
  };

  return (
    <>
      <LoaderWrapper isLoading={loading} loaderText="">
        <Formik
          initialValues={{ newEmploymentStatusType: "" }}
          validationSchema={schema}
          validateOnChange={false}
          validateOnBlur={false}
          onSubmit={onFormSubmit}
        >
          {({ handleChange, handleSubmit, values }) => {
            return (
              <>
                <div className="dropdside-custom-dropdown">
                  <div>
                    <SearchableList
                      label="Select an Employment Status or create a new one"
                      options={data}
                      onSelect={onSelect}
                      selectedValue={value}
                    />
                  </div>
                  <div className="line-separator"></div>
                  <Form autoComplete="off" onSubmit={handleSubmit}>
                    <Field
                      label="Add a new Employment Status type :"
                      name={"newEmploymentStatusType"}
                      type="text"
                      value={values["newEmploymentStatusType"]}
                      onChange={handleChange}
                      autoComplete="nope"
                      component={FieldInput}
                      placeholder={"Employee status"}
                      onKeyDown={keyDown}
                    />
                    <Form.Group>
                      <Button type="submit">Add</Button>
                    </Form.Group>
                  </Form>
                </div>
              </>
            );
          }}
        </Formik>
      </LoaderWrapper>
    </>
  );
};

export default EmploymentStatusDropdown;
