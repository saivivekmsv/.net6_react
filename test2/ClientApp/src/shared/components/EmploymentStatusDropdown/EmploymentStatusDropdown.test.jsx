import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import EmploymentStatusDropdown from "./EmploymentStatusDropdown";
import employmentStatus from "../../mocks/employmentStatus.json";

describe("EmploymentStatusDropdown", () => {
  const props = {};
  test("to match snapshot", () => {
    const { baseElement } = render(
      <Router>
        <EmploymentStatusDropdown data={employmentStatus.data} {...props} />
      </Router>
    );
    expect(baseElement).toMatchSnapshot();
  });
});
