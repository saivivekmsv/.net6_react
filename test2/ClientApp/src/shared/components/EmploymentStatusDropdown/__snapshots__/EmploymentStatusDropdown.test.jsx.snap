// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmploymentStatusDropdown to match snapshot 1`] = `
<body>
  <div>
    <div
      class="dropdside-custom-dropdown"
    >
      <div>
        <div
          class="searchable-list-input form-group"
        >
          <label
            class="searchable-list-label form-label"
          >
            Select an Employment Status or create a new one
          </label>
        </div>
        <div
          class="searchable-list overflow-y-auto"
        >
          <button
            class="dropdown-item"
            data-attr="1"
            type="button"
          >
            Active
          </button>
          <button
            class="dropdown-item"
            data-attr="2"
            type="button"
          >
            Retired
          </button>
          <button
            class="dropdown-item"
            data-attr="3"
            type="button"
          >
            Terminated
          </button>
          <button
            class="dropdown-item"
            data-attr="4"
            type="button"
          >
            Leave
          </button>
          <button
            class="dropdown-item"
            data-attr="5"
            type="button"
          >
            Military Leave
          </button>
          <button
            class="dropdown-item"
            data-attr="6"
            type="button"
          >
            Deceased
          </button>
          <button
            class="dropdown-item"
            data-attr="7"
            type="button"
          >
            Disabled
          </button>
          <button
            class="dropdown-item"
            data-attr="8"
            type="button"
          >
            Others
          </button>
        </div>
      </div>
      <div
        class="line-separator"
      />
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Add a new status name
        </label>
        <input
          class="form-control"
          name="newEmploymentStatus"
          type="text"
        />
      </div>
      <div
        class="form-group"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          Add
        </button>
      </div>
    </div>
  </div>
</body>
`;
